### PR TITLE
enhance: Add restful api for devops to execute rolling upgrade

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #29261
This PR Add restful api for devops to execute rolling upgrade,  including suspend/resume balance and manual transfer segments/channels.

### CODE REVIEW

**Comment by weiliu1031** [comment]
rerun ut

**Comment by weiliu1031** [comment]
rerun ut

**Comment by weiliu1031** [comment]
rerun ut

**Comment by weiliu1031** [comment]
rerun ut

File: internal/querycoordv2/handlers.go [hunk 1]
```diff
@@ -87,86 +88,150 @@ func (s *Server) getCollectionSegmentInfo(collection int64) []*querypb.SegmentIn
...
-		}
+// generate balance segment task and submit to scheduler
+func (s *Server) balanceSegments(ctx context.Context,
+	collectionID int64,
+	replicaID int64,
+	srcNode int64,
+	dstNodes []int64,
+	segments []*meta.Segment,
+	sync bool,
+	copyMode bool,
```
Comment (yiwangdr) [inline internal/querycoordv2/handlers.go:99]: what is the copyMode field? Could you add some comments here?

File: internal/querycoordv2/services.go
[the remark this answers is not in the document]
Comment (weiliu1031) [reply internal/querycoordv2/services.go:1]: sure, we have `TestLoadBalance` in `services_test` to cover `LoadBalance`'s behavior

File: internal/querycoordv2/services.go [hunk 2]
```diff
@@ -681,24 +681,68 @@ func (s *Server) LoadBalance(ctx context.Context, req *querypb.LoadBalanceReques
...
 		}
+	}
+
+	// check whether dstNode is healthy
+	for dstNode := range dstNodeSet {
 		if err := s.isStoppingNode(dstNode); err != nil {
 			return merr.Status(errors.Wrap(err,
 				fmt.Sprintf("can't balance, because the destination node[%d] is invalid", dstNode))), nil
 		}
+		dstNodeSet.Insert(dstNode)
```
[the remark this answers is not in the document]
Comment (weiliu1031) [reply internal/querycoordv2/services.go:708]: duplicate code logic, should be removed

**Comment by weiliu1031** [comment]
rerun ut

File: internal/querycoordv2/services.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiwangdr) [reply internal/querycoordv2/services.go:709]: Wait, it's still not deleted. You probably didn't commit/push your change.

File: internal/querycoordv2/balance/balance.go [hunk 3]
```diff
@@ -67,7 +69,18 @@ type RoundRobinBalancer struct {
...
-func (b *RoundRobinBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64) []SegmentAssignPlan {
+func (b *RoundRobinBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64, manualBalance bool) []SegmentAssignPlan {
+	// skip out suspend node and stopping node during assignment, but skip this check for manual balance
+	if !manualBalance {
+		nodes = lo.Filter(nodes, func(node int64, _ int) bool {
+			info := b.nodeManager.Get(node)
+			if info != nil && info.GetState() == session.NodeStateNormal {
+				return true
+			}
+			return false
```
Comment (yiwangdr) [inline internal/querycoordv2/balance/balance.go:80]: we can simplify it with
`return info != nil && info.GetState() == session.NodeStateNormal`

File: internal/querycoordv2/balance/balance.go [hunk 4]
```diff
@@ -90,7 +103,17 @@ func (b *RoundRobinBalancer) AssignSegment(collectionID int64, segments []*meta.
...
-func (b *RoundRobinBalancer) AssignChannel(channels []*meta.DmChannel, nodes []int64) []ChannelAssignPlan {
+func (b *RoundRobinBalancer) AssignChannel(channels []*meta.DmChannel, nodes []int64, manualBalance bool) []ChannelAssignPlan {
+	// skip out suspend node and stopping node during assignment, but skip this check for manual balance
+	if !manualBalance {
+		nodes = lo.Filter(nodes, func(node int64, _ int) bool {
+			info := b.nodeManager.Get(node)
+			if info != nil && info.GetState() == session.NodeStateNormal {
+				return true
+			}
+			return false
```
Comment (yiwangdr) [inline internal/querycoordv2/balance/balance.go:114]: ditto

File: internal/querycoordv2/balance/rowcount_based_balancer.go [hunk 5]
```diff
@@ -41,7 +41,18 @@ type RowCountBasedBalancer struct {
...
-func (b *RowCountBasedBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64) []SegmentAssignPlan {
+func (b *RowCountBasedBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64, manualBalance bool) []SegmentAssignPlan {
+	// skip out suspend node and stopping node during assignment, but skip this check for manual balance
+	if !manualBalance {
+		nodes = lo.Filter(nodes, func(node int64, _ int) bool {
+			info := b.nodeManager.Get(node)
+			if info != nil && info.GetState() == session.NodeStateNormal {
+				return true
+			}
+			return false
```
Comment (yiwangdr) [inline internal/querycoordv2/balance/rowcount_based_balancer.go:52]: ditto

File: internal/querycoordv2/balance/rowcount_based_balancer.go [hunk 6]
```diff
@@ -75,7 +86,18 @@ func (b *RowCountBasedBalancer) AssignSegment(collectionID int64, segments []*me
...
-func (b *RowCountBasedBalancer) AssignChannel(channels []*meta.DmChannel, nodes []int64) []ChannelAssignPlan {
+func (b *RowCountBasedBalancer) AssignChannel(channels []*meta.DmChannel, nodes []int64, manualBalance bool) []ChannelAssignPlan {
+	// skip out suspend node and stopping node during assignment, but skip this check for manual balance
+	if !manualBalance {
+		nodes = lo.Filter(nodes, func(node int64, _ int) bool {
+			info := b.nodeManager.Get(node)
+			if info != nil && info.GetState() == session.NodeStateNormal {
+				return true
+			}
+			return false
```
Comment (yiwangdr) [inline internal/querycoordv2/balance/rowcount_based_balancer.go:97]: ditto

File: internal/querycoordv2/balance/score_based_balancer.go [hunk 7]
```diff
@@ -50,7 +50,18 @@ func NewScoreBasedBalancer(scheduler task.Scheduler,
...
-func (b *ScoreBasedBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64) []SegmentAssignPlan {
+func (b *ScoreBasedBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64, manualBalance bool) []SegmentAssignPlan {
+	// skip out suspend node and stopping node during assignment, but skip this check for manual balance
+	if !manualBalance {
+		nodes = lo.Filter(nodes, func(node int64, _ int) bool {
+			info := b.nodeManager.Get(node)
+			if info != nil && info.GetState() == session.NodeStateNormal {
+				return true
+			}
+			return false
```
Comment (yiwangdr) [inline internal/querycoordv2/balance/score_based_balancer.go:61]: ditto

File: internal/querycoordv2/services.go [hunk 8]
```diff
@@ -682,24 +682,68 @@ func (s *Server) LoadBalance(ctx context.Context, req *querypb.LoadBalanceReques
...
-			err := merr.WrapErrNodeNotFound(dstNode, "destination node not found in the same replica")
-			log.Warn("failed to balance to the destination node", zap.Error(err))
-			return merr.Status(err), nil
+
+	// when no dst node specified, default to use all other nodes in same
+	dstNodeSet := typeutil.NewUniqueSet()
+	if len(req.GetDstNodeIDs()) == 0 {
+		outboundNodes := s.meta.ResourceManager.CheckOutboundNodes(replica)
+		availableNodes := lo.Filter(replica.Replica.GetNodes(), func(node int64, _ int) bool { return !outboundNodes.Contain(node) })
+		dstNodeSet.Insert(availableNodes...)
```
Comment (yiwangdr) [inline internal/querycoordv2/services.go:691]: Need testing coverage

File: internal/querycoordv2/services.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiwangdr) [reply internal/querycoordv2/services.go:709]: Wait, it's still not deleted. You probably didn't commit/push your change.

**Review by yiwangdr — COMMENTED** [review]
It looks great! Thanks for adding the integration tests.

**Review by yiwangdr — APPROVED** [review]

File: internal/querycoordv2/session/node_manager.go [hunk 9]
```diff
@@ -61,6 +67,36 @@ func (m *NodeManager) Stopping(nodeID int64) {
...
 }
 
+func (m *NodeManager) Suspend(nodeID int64) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	nodeInfo, ok := m.nodes[nodeID]
+	if !ok {
+		return merr.WrapErrNodeNotFound(nodeID)
+	}
+	nodeInfo.SetState(NodeStateSuspend)
```
[the remark this answers is not in the document]
Comment (weiliu1031) [reply internal/querycoordv2/session/node_manager.go]: add new constraints here: only normal node can be suspended,and only suspended node can be resumed

**Comment by yiwangdr** [comment]
@czs007 @congqixia @xiaofan-luan friendly ping

**Comment by yiwangdr** [comment]
@czs007 @congqixia @xiaofan-luan friendly ping again. We must have it in 2.4

**Comment by xiaofan-luan** [comment]
why no directly use k8s operators?

**Comment by xiaofan-luan** [comment]
> @czs007 @congqixia @xiaofan-luan friendly ping again. We must have it in 2.4

anyone is blocking on this?
I thought we did some major refactor on the resource group interface

**Comment by yiwangdr** [comment]
> > @czs007 @congqixia @xiaofan-luan friendly ping again. We must have it in 2.4
> 
> anyone is blocking on this? I thought we did some major refactor on the resource group interface

No, we need one of the owners to merge it.

**Comment by weiliu1031** [comment]
rerun ut

**Comment by weiliu1031** [comment]
rerun ut

**Pull request merged**